add data-modeling docs (#6657)
microsoft/FluidFramework: docs/content/docs/build/containers.md

menuPosition: 2
---

## Overview

> The container is the primary unit of encapsulation in Fluid. It consists of a collection of Loadable Objects and supporting lifecycle APIs. The container is represented as the `FluidContainer` type.

Author: I'm going to commit this for now while be bottom out on the Loadable Object conversation. I love your edits by the way 💯

The below docs are broken in to `Creation & loading` and `Interaction & lifecycle`.

## Creating & loading

Creation and loading happens from specific `[service]-client` packages. Each `[service]-client` package contains a common API structure around container interaction. The common API structure is around defining the container schema and getting the `FluidContainer` object.

For the below scenarios we will be using the `client` to represent the `[service]-client` portion. For specific `[service]-client` usage see their corresponding documentation.

### Container schema

> The container schema defines the data that the container instance knows about. It includes `initialObjects` that will synchronously loaded and always available. It also includes a definition of types that can be dynamically created. The same schema definition must be provided for both. For more information on `initialObjects` and dynamically cration see [Data modeling](./data-modeling.md)

Author: ```suggestion
You must define a schema that represents the structure of the data within your container. A schema can include `initialObjects` that are always available and types that can be dynamically created by the container at runtime. The same schema definition must be provided for creation and subsequent loading of the container. For more information on `initialObjects` and dynamic object creation see [Data modeling](./data-modeling.md).
```

> The container schema defines the data that the container instance knows about. It includes `initialObjects` that will synchronously loaded and always available. It also includes a definition of types that can be dynamically created. The same schema definition must be provided for both. For more information on `initialObjects` and dynamically cration see [Data modeling](./data-modeling.md)

Reviewer: Your suggestion looks good to me!

```typescript
const schema = {
    name: "example-container",
    initialObjects: {
        layout: SharedDirectory,
        text: SharedString
    },
    dynamicObjectTypes: [ SharedCell, SharedString ],
}

> ```

Author: @tylerbutler, how do you read the word `or` in the above `creation *or* load` context. I read it as exclusive IE. it could be available when you create or when you load. But I think you mean it like. It is available when you create or when you load.

> ```

Reviewer: No, I didn't mean it in the exclusive sense. Agree it's confusing. Maybe "which will be available immediately when the container is loaded?" In this case, the difference between loaded and created isn't relevant.

### How to create a container

Containers are created off the `client` instance via the `createContainer` function. This function always takes a config that is specific to the service and a schema object that defines the container.

```typescript
const schema = {
    name: "example-container",
    initialObjects: {
        layout: SharedDirectory,
        text: SharedString
    },
    dynamicObjectTypes: [ SharedCell, SharedString ],
}
const { container, containerServices} = await client.createContainer(/*service config*/, schema);

> ```

Author: You can highlight! 
![MagicShiaLabeoufGIF](https://user-images.githubusercontent.com/6289462/125137355-5f78e300-e0c1-11eb-89bd-cf4bafd9c946.gif)

### How to load a container

> To load the container we created in the above section we must provide the service config as well as the exact same schema definition. The same container schema is required on all subsequent loads to ensure the container is loaded correctly.

Author: Good callout I should add an explicit section. It depends on what you change but the outcome is almost always bad. Basically it's all the problems around schema versioning.

If you change the initialObjects it will just blow up when it's trying to load them. If you add a new there's no creation flow. If you remove on it will try to load it and fail.

If you remove a dynamicType and that type is being used it will blow up when trying to load that type.

If you add a new dynamicType then use it and another user doesn't have that type they will blow up.

```typescript
const schema = {
    name: "example-container",
    initialObjects: {
        layout: SharedDirectory,
        text: SharedString
    },
    dynamicObjectTypes: [ SharedCell, SharedString ],
}
const { container, containerServices} = await client.getContainer(/*service config*/, schema);
```

### Attaching a container

Once the `createContainer` call resolves the container is connected to the service (attached) and available to collaborative. 

> We are currently working on an advanced scenario where containers can be created locally before persisting them to the server. The primary use case for this is if a client want's to draft the initial state of a container before other collaborators enter.

Author: After review we decided to not include non-supported content.

> 
> We are currently working on an advanced scenario where containers can be created locally before persisting them to the server. The primary use case for this is if a client want's to draft the initial state of a container before other collaborators enter.

Author: ```suggestion
```

});

container.on("connected", () => {
    // handle connected
    // enable editing if disabled
});
```

### dispose

The container exposes functionality that allow cleanup of the container object. Calling dispose will remove any server connections and clean up registered events. Once a container is dispose it cannot be re-hydrated. Connecting to the same container requires another `getContainer(...)` call.

```typescript
container.dispose();

const disposed = container.disposed();

container.on("disposed", () => {
    // handle event cleanup to prevent memory leaks
});
```

### initialObjects

`initialObjects` are the base set of `LoadableObjects` in a container. They exist for the lifetime of the container and are guaranteed to be available and loadable. They are defined via the container schema and match the signature on the schema. 

For more on `initialObjects` see [Data modeling](data-modeling.md)

### create

The container also exposes a create function that allows dynamic `LoadableObject` creation.

For more on dynamic object creation see [Data modeling](data-modeling.md)

## Patterns for managing container lifecycle

### Create/load separation

> When creating and loading a container our basic guidance is to separate the two flows. This provides a cleaner separation of responsibilities within the code it self. From a scenario perspective it manifests as the creator going through a explicit creation process that results in a redirect to a new page who's sole responsibility is to load the container. All users will load the container through this subsequent flow.

Author: I'm not sure I really like this section since it's a bit to prescriptive. I do like that it highlights creation though so probably worth keeping for now.

It should be noted that the downside of this approach is that the connection needs to be established twice. Once for the creation and once for the load. Depending on your scenario you should choose the pattern that works best for you.

### Multi-container example

Multiple containers can be loaded from an app, or on a page, at the same time. There are two primary scenarios where you application would be using multiple containers.

First, is if your application loads two different experiences that have different underlying data structures. `Experience 1` may require a `SharedMap` and `Experience 2` may requires a `SharedString`. To minimize our memory footprint we can create two different container schemas and load only what we need. In this scenario our app can load two different containers but we are only choosing to load one at a time.

> The second scenario involves loading two containers at once. Currently, all services enable permissioning at the container level, so we can use containers as a natural boundary for restricting access. An example scenario for this would be building an education app where mutliple teachers are gathering with students. The students and teachers may have a shared view while the teachers may also want to have an additional private view on the side. In this scenario the students would be loading one container and the teachers would be loading two.

Author: The only thing I don't like is stating that containers serve as a auth boundary. Containers can be used as a auth boundary but it's dependent on the service to support it and the developer to enable it. If the user doesn't auth correctly at the container level there's nothing we can do about it.

I used the word permissioning because you need specific a token for each container you are creating/loading. In this sense we are implying a potential auth boundary but not stating it. 

I would be much more comfortable talking about auth in the context of the frs docs because we can explicitly say if you implement this specific auth flow you can achieve this scenario. But as the generic client platform we can't make such garuantees.

    name: "example-container",
    initialObjects: {
        layout: SharedDirectory,
        text: SharedString
    },
    dynamicObjectTypes: [ SharedCell, SharedString ],
}
```

### Creating a container

Containers are created from the service-specific client's `createContainer` function. You must provide a config that is specific to the service and a schema object that defines the container schema.

```typescript {hl_lines=[10]}
const schema = {
    name: "example-container",
    initialObjects: {
        layout: SharedMap,
    },
}
const { container, containerServices} =
    await client.createContainer(/*service config*/, schema);
```

### Loading a container

To load the container created in the above section you must provide the service config as well as the exact same schema definition. The same container schema is required on all subsequent loads or the container will not be loaded correctly.

```typescript {hl_lines=[10]}
const schema = {
    name: "example-container",
    initialObjects: {
        layout: SharedMap,
    },
}
const { container, containerServices} =
    await client.getContainer(/*service config*/, schema);
```

### Attaching a container

> Once the `createContainer` or `loadContainer` function calls complete, the returned container is _attached_ -- that is, it  is connected to the Fluid service -- and ready to power collaboration. 

Author: ```suggestion
Once the `createContainer` or `loadContainer` function calls complete, the returned container is _attached_ -- that is, it is connected to the Fluid service -- and ready to power collaboration. 
```

```typescript
container.dispose();

const disposed = container.disposed();

container.on("disposed", () => {
    // handle event cleanup to prevent memory leaks
});
```

### initialObjects

`initialObjects` are shared objects that you define in a container's schema and exist for the lifetime of the container. These shared objects are exposed via the `initialObjects` property on the container.

```typescript
const schema = {
    name: "example-container",
    initialObjects: {
        layout: SharedMap,
        text: SharedString
    },
}

// ...

const layout = container.initialObjects.layout;
const text = container.initialObjects.text;
```

For more information about `initialObjects` see [Data modeling](data-modeling.md).

### create

The container also exposes a create function that allows dynamic creation of shared objects. This enables containers to create Fluid objects dynamically at runtime.

For more information about dynamic object creation see [Data modeling](data-modeling.md).

## Patterns for managing container lifecycle

### Create/load separation

> When creating and loading a container, it can be tempting to have a consistent code path for both creation and loading. However, in Fluid it is generally better to separate the two flows. This provides a cleaner separation of responsibilities within the code itself. Also, in typical use-cases, a user will create a new container through some UI action that results in a redirect to another page whose sole responsibility is to load a container. All subsequent users will load the container by navigating directly to that page.

Author: ```suggestion
When creating and loading a container, it can be tempting to have a consistent code path for both creation and loading. However, it is generally recommended to separate the two flows. This provides a cleaner separation of responsibilities within the code itself. Also, in typical use-cases, a user will create a new container through some UI action that results in a redirect to another page whose sole responsibility is to load a container. All subsequent users will load the container by navigating directly to that page.
```

menuPosition: 2
---

## Overview

The container is the primary unit of encapsulation in the Fluid Framework. A container is represented by the `FluidContainer` type and consists of a collection of shared objects and APIs to manage the lifecyle of those objects.

> This documentation will explain how to create and load containers, the APIs to interact with them, and the overall container lifecycle.```

Author: ```suggestion
This documentation will explain how to create and load containers, the APIs to interact with them, and the overall container lifecycle.
```

## Creating & loading

You can create containers using APIs provided by a service-specific client package. Each service-specific client package contains a common API for manipulating containers. For example, the `tinylicious-client` package provides these APIs for the Tinylicious Fluid service. These common APIs enable you to define the container schema and retrieve the `FluidContainer` object.

In the below scenarios, `client` represents the service-specific client. See the documentation for the service you are using for more details about how to use the service-specific client.

### Container schema

You must define a schema that represents the structure of the data within your container. A schema can include `initialObjects` that are always available and types that can be dynamically created by the container at runtime. The same schema definition must be provided for creation and subsequent loading of the container. For more information on `initialObjects` and dynamic object creation see [Data modeling](./data-modeling.md).

> This example schema defines two initial objects, and also enables the container to create additional `SharedCell`s and `SharedString`s at runtime.

Author: ```suggestion
This example schema defines two initial objects, and declares `SharedCell` and `SharedString` as shared object types that can be dynamically created at runtime.
```

```typescript
const schema = {
    name: "example-container",
    initialObjects: {
        layout: SharedDirectory,
        text: SharedString
    },
    dynamicObjectTypes: [ SharedCell, SharedString ],
}
```

> ### How to create a container

Author: ```suggestion
### Creating a container
```

Containers are created from the service-specific client's `createContainer` function. You must provide a config that is specific to the service and a schema object that defines the container schema.

```typescript {hl_lines=[10]}

> const schema = {
>     name: "example-container",
>     initialObjects: {
>         layout: SharedDirectory,
>         text: SharedString
>     },
>     dynamicObjectTypes: [ SharedCell, SharedString ],
> }

Author: ```suggestion
const schema = {
    name: "example-container",
    initialObjects: {
        layout: SharedMap,
    },
}
```

const { container, containerServices} =
    await client.createContainer(/*service config*/, schema);
```

### Loading a container

To load the container created in the above section you must provide the service config as well as the exact same schema definition. The same container schema is required on all subsequent loads or the container will not be loaded correctly.

```typescript {hl_lines=[10]}

> const schema = {
>     name: "example-container",
>     initialObjects: {
>         layout: SharedDirectory,
>         text: SharedString
>     },
>     dynamicObjectTypes: [ SharedCell, SharedString ],
> }

Author: ```suggestion
const schema = {
    name: "example-container",
    initialObjects: {
        layout: SharedMap,
    },
}
```
